main/src/main/java/sbt/internal/BottomClassLoader.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * sbt
+ * Copyright 2011 - 2018, Lightbend, Inc.
+ * Copyright 2008 - 2010, Mark Harrah
+ * Licensed under Apache License 2.0 (see LICENSE)
+ */
+
+package sbt.internal;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URL;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import sbt.util.Logger;
+
+/**
+ * The bottom most layer in our layering hierarchy. This layer should never be cached. The
+ * dependency layer may need access to classes only available at this layer using java reflection.
+ * To make this work, we register this loader with the parent in its constructor. We also add the
+ * lookupClass method which gives ReverseLookupClassLoader a public interface to findClass.
+ *
+ * <p>To improve performance, when loading classes from the parent, we call the loadClass method
+ * with the reverseLookup flag set to false. This prevents the ReverseLookupClassLoader from trying
+ * to call back into this loader when it can't find a particular class.
+ */
+final class BottomClassLoader extends ManagedClassLoader {
+  private final ReverseLookupClassLoaderHolder holder;
+  private final ReverseLookupClassLoader parent;
+  private final Set<String> lookedUp = ConcurrentHashMap.newKeySet();
+  private final ClassLoadingLock classLoadingLock = new ClassLoadingLock();
+
+  BottomClassLoader(
+      final ReverseLookupClassLoaderHolder holder,
+      final URL[] dynamicClasspath,
+      final ReverseLookupClassLoader reverseLookupClassLoader,
+      final File tempDir,
+      final boolean allowZombies,
+      final Logger logger) {
+    super(dynamicClasspath, reverseLookupClassLoader, allowZombies, logger);
+    setTempDir(tempDir);
+    this.holder = holder;
+    this.parent = reverseLookupClassLoader;
+    parent.setDescendant(this);
+  }
+
+  static {
+    ClassLoader.registerAsParallelCapable();
+  }
+
+  final Class<?> reverseLoad(final String name, final boolean resolve)
```

Comment:
What does reverseLoad mean?

Reply:
Since loading is `String => Class`, a reverse would be `Class => String`.

Is `loadOnceForParent(String, Boolean): Class` what's going on?

Reply:
I chose reverseLoad to mean that we were traversing the classloading hierarchy in the _reverse_ direction. Normally a child loads a class from it’s parent but here the parent tries to load a class from a child.

Reply:
Right, and the `BoomClassLoader` is the child who gets sacrificed by the parent right? So I think `loadOnce` or `loadForParent` might be a better name with comment saying it gets called by the parent.

Also instead of reversing, would it make sense to have two parents, like a primary parent, and secondary parent? If the primary fails for whatever reason lazily create a flat `ClassLoader`?

Reply:
Instead of renaming `reverseLoad`, I just removed it and made `findClass` public in `BottomClassLoader`.

Reply:
> Also instead of reversing, would it make sense to have two parents, like a primary parent, and secondary parent? If the primary fails for whatever reason lazily create a flat ClassLoader?

I don't think I understand this idea. Is the idea that we'd create a dual loader where the primary classloader has the dependency classpath and the secondary loader has the full classpath?

Reply:
Yea. So in that world, there's no traversing back to child, it would just fall back a flat secondary parent.

Reply:
That would not work because the classes loaded by the secondary parent cannot be used by the child.

Reply:
What I've been trying to do is isolate things so that there is at most one classloader that can load each class on the classpath (this could be violated if multipe jars define the same class). Dual loaders make it possible for multiple classloaders to load the same class which makes it much easier for something to go wrong.

Reply:
To make it concrete, the java-serialization scripted test defines an object that looks like this:
```
object Reflection {
  def roundTrip[A](a: A): A = {
    val baos = new ByteArrayOutputStream()
    val oos = new ObjectOutputStream(baos)
    oos.writeObject(a)
    oos.close()
    val bais = new ByteArrayInputStream(baos.toByteArray())
    val ois = new ObjectInputStream(bais)
    try ois.readObject().asInstanceOf[A]
    finally ois.close()
  }
}
```
The call to `ois.readObject` will eventually lead to a call to `Class.forName` with the class name read from the object stream. Suppose that we invoke this method with a class `A` that is defined in the project classpath. If there were a secondary flatloader that were used to load `A`, then the `asInstanceOf[A]` would cause a `ClassCastException`. By looking back into the child, however, we load the same version of `A` that was passed in as input.

Reply:
> Loading those classes in alternate secondary parent loader creates an alternate classloading hierarchy which would be incompatible with the normal hierarchy and would almost surely cause frequent incompatible class exceptions.

Yea. That makes sense.